GH-520 clear various caches before updating references

JanKoehnlein opened the pull request.

cdietrich commented on org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/RenameIssue520Test.xtend:
```diff
@@ -1,47 +0,0 @@
-/*******************************************************************************
```
why delete the test?

cdietrich replied:
where the other expectations wrong?

JanKoehnlein replied:
It's duplicate. Please see the comment below.

JanKoehnlein replied:
Yes, the expectations were bogus.

cdietrich replied:
ahh yes did not see that the other test actually tested the wrong thing.

JanKoehnlein commented:
There are two usecases: In Eclipse and on the language server.

In both cases, we have to clear the resource cache of related resources before updating the cross references. The cache holds the scopes and the visible containers in the state before the modification has been applied. It has not been cleared on the related resource because we haven't applied a change to it yet.

The language server attaches the ChunkedResourceDescriptions to the resource set. We have to remove them as these again reflect the state before the modification. We have to add them after processing the resource because further related resources may be loaded in the following, and may rely on this state.

I've also removed the test because it is equivalent to `RenameTest.testRenameOnReferenceInOtherFile()`

cdietrich commented:
LGTM

JanKoehnlein commented:
After some discussion I dropped the idea of just clearing caches in specific usecases. It does not appear to be sustainable. 

The main problem is that we don't have the notion of a live scope in IDE as we have in UI, i.e. that the current contents of a resource set shadow the 'persisted' state for scoping. The latest commit introduces that.

Note that we bypass the entire IAllContainerState thingy, as in this scenario the entire index is attached as ChunkedResourceDescriptions to the resource set, i.e. there is no global state as in the Eclipse builder scenario.

This is just a proposal for discussion. The final shadowing strategy could stand some more love, e.g. caching, real shadowing (local descriptions are just put before the persisted ones in the lookup list), more tests. It'd also be interesting how we could deal with container changes, i.e. when a file is moved from one project to another (we don't have that for Eclipse either).

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedAllContainerState.java:
```diff
@@ -19,10 +19,10 @@
 
 /**
- * This class ensures that every IResourceDescription from a local IResourceDescrions that is not known to the global
+ * This class ensures that every IResourceDescription from a local IResourceDescrptions that is not known to the global
```
Still a typo in the comment

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedAllContainerState.java:
```diff
@@ -19,10 +19,10 @@
  * IAllContainerState gets a proper container.
  * 
- * This is usually necessary when an IResourceDescrion has no (not yet a) file in the file system.
+ * This is usually necessary when an IResourceDescrption has no (not yet a) file in the file system.
```
Same here

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedAllContainerState.java:
```diff
@@ -31,7 +31,7 @@
 	public static class Provider {
 		// we're not implementing IAllContainerState.Provider here because its get() expects a global instance of IResourceDescrions. 
-		// Here, we expect one that only holds an IResourceDescrion for every resource from a ResourceSet.  
+		// Here, we expect one that only holds an IResourceDescripton for every resource from a ResourceSet.  
```
and here

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainer.xtend:
```diff
@@ -0,0 +1,94 @@
+	protected def boolean isContained(URI uri) {
+		chunk
+			.allResourceDescriptions
+			.exists[URI == uri]
```
Is a linear search the best we have here?

szarnekow commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService.xtend:
```diff
@@ -47,19 +48,22 @@ class RenameService implements IRenameService {
 			if (xtextResource instanceof XtextResource) {
-				val element = xtextResource.resolveElementAt(offset)
-				if (element === null)
+				val element = EcoreUtil.resolve(xtextResource.resolveElementAt(offset), resourceSet)
```
`resolveElementAt` should yield a resolved EObject. Why is the failsafe seemingly redundant call here necessary?

JanKoehnlein commented:
I've fixed the typos, addressed the complexity problem and added real shadowing and tests. 
I also use the IProjectConfig if available to determine the container of files that are new.
Please review.

cdietrich commented on org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/RenameIssue520Test.xtend:
```diff
@@ -1,47 +0,0 @@
- */
- // FIXME https://github.com/eclipse/xtext-core/issues/520
- @Ignore("https://github.com/eclipse/xtext-core/issues/520")
-class RenameIssue520Test extends AbstractTestLangLanguageServerTest {
```
miss the delete for the generated file

JanKoehnlein replied:
done

cdietrich commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/DefaultIdeModule.xtend:
```diff
@@ -20,4 +28,17 @@ class DefaultIdeModule extends AbstractGenericModule {
+	}
+	
+	def Class<? extends IContainer.Manager> bindIContainer$Manager() {
+		ProjectDescriptionBasedContainerManager
```
can this break existing clients that expect SimpleResourceDescriptionsBasedContainerManager?

JanKoehnlein replied:
Good catch. It delegates anyway. Removed binding

cdietrich commented on org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/DefaultIdeModule.xtend:
```diff
@@ -20,4 +28,17 @@ class DefaultIdeModule extends AbstractGenericModule {
+	}
+	
+	def Class<? extends IProjectConfigProvider> bindIProjectConfigProvider() {
+		ProjectConfigProvider
```
why do we use this in XbaseRuntimeModule but not DefaultRuntimeModule?

cdietrich replied:
No I means the binding is inconsistent done for xbase and non xbase plus defaultide 

So I wonder if it should go to defaultruntime

JanKoehnlein replied:
You mean in DefautIdeRuntimeModule but not DefaultRuntimeModule?

We could. But this will definitely affect existing clients and I am not sure about the impact.

cdietrich replied:
No I mean it’s inconsistent

cdietrich replied:
Why is binding in xbaseruntimemodule

cdietrich commented on org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/resource/containers/SourceFolder.java:
```diff
@@ -0,0 +1,37 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.xtext.resource.containers;
```
maybe move to testing or even to xtext core?

JanKoehnlein replied:
Maybe when we need it again...

cdietrich commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/ProjectDescriptionBasedContainerManager.xtend:
```diff
@@ -25,38 +26,48 @@ import org.eclipse.xtext.resource.impl.ResourceDescriptionsData
+		
+		val resourceSet = chunkedResourceDescriptions.resourceSet
+		val projectDescription = ProjectDescription.findInEmfObject(resourceSet)
+		val container = createContainer(resourceDescriptions, chunkedResourceDescriptions, projectDescription.name)
```
null safe?

JanKoehnlein replied:
Guarded by `shouldUseProjectDescriptionBasedContainers()`

cdietrich commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/impl/LiveShadowedResourceDescriptions.java:
```diff
@@ -43,15 +44,18 @@
 	
 	@Inject
-	private ResourceSetBasedResourceDescriptions localDescriptions;
+	protected ResourceSetBasedResourceDescriptions localDescriptions;
```
why no use getter?

JanKoehnlein replied:
Because we set it in the new subclass.

cdietrich commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/impl/LiveShadowedResourceDescriptions.java:
```diff
@@ -43,15 +44,18 @@
-	private IResourceDescriptions globalDescriptions;
+	private Provider<IResourceDescriptions> globalDescriptionsProvider;
 
+	protected IResourceDescriptions globalDescriptions;
```
same here

JanKoehnlein replied:
Because we set it in the new subclass.

cdietrich replied:
and whats the setter for?

szarnekow replied:
Would it make sense to have a setter like method that accepts both values as arguments rather than breaking the encapsulation?

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/ProjectDescriptionBasedContainerManager.xtend:
```diff
@@ -25,38 +26,48 @@ import org.eclipse.xtext.resource.impl.ResourceDescriptionsData
+		val resourceSet = chunkedResourceDescriptions.resourceSet
+		val projectDescription = ProjectDescription.findInEmfObject(resourceSet)
+		val allContainers = <IContainer>newArrayList
+		allContainers.add(createContainer(resourceDescriptions, chunkedResourceDescriptions, projectDescription.name))
```
null safety here, too?

JanKoehnlein replied:
See comment above

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/ProjectDescriptionBasedContainerManager.xtend:
```diff
@@ -25,38 +26,48 @@ import org.eclipse.xtext.resource.impl.ResourceDescriptionsData
+		val projectDescription = ProjectDescription.findInEmfObject(resourceSet)
+		val allContainers = <IContainer>newArrayList
+		allContainers.add(createContainer(resourceDescriptions, chunkedResourceDescriptions, projectDescription.name))
+		for (name : projectDescription.dependencies) {
```
and here

JanKoehnlein replied:
See comment above

szarnekow commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/impl/LiveShadowedResourceDescriptions.java:
```diff
@@ -43,15 +44,18 @@
 public class LiveShadowedResourceDescriptions extends ResourceSetBasedResourceDescriptions {
```
Please also implement IShadowedResourceDescriptions

JanKoehnlein commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainer.xtend:
```diff
@@ -0,0 +1,122 @@
+
+	protected def getProjectConfig() {
+		if (!isProjectConfigSet) {
+			projectConfig = descriptions.workspaceConfig?.projects?.findFirst[name == containerName]
```
done

JanKoehnlein commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainer.xtend:
```diff
@@ -0,0 +1,122 @@
+	}
+	
+	protected def boolean isContained(URI uri) {
+		chunk.getResourceDescription(uri) !== null || getProjectConfig?.sourceFolders?.exists[path.isPrefixOf(uri)]
```
done

JanKoehnlein commented on org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainer.xtend:
```diff
@@ -0,0 +1,122 @@
+	}
+	
+	override hasResourceDescription(URI uri) {
+		chunkedResourceDescriptions.getContainer(containerName).getResourceDescription(uri) !== null
```
done

JanKoehnlein commented:
I squashed the corrections due to your comments. Merging now.
Let's discuss the DI module issue separately

szarnekow commented on org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/containers/LiveShadowedChunkedContainerTest.xtend:
```diff
@@ -0,0 +1,169 @@
+	}
+	
+	@Test
+	def testDeleteElement() {
```
Could you please add a test for a resource rename and resource deletion?
